🚧 feat: add support for vmrest

niwamo opened the pull request:
As described in #157:

Creates a new driver for vmrest, the API that ships with VMWare Workstation Pro. Due to limitations in the API, only the vmware-vmx builder can be implemented. 

Checks have been added to the vmware-iso builder to throw an explanatory error if a user mistakenly tries to use the vmrest remote_type. Additional checks have been added to throw errors if incompatible configurations are set while using the vmrest remote_type. 

The vmware-vmx documentation was also updated. 

A test for the new driver has been provided in #157.

Closes #157

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer-plugin-vmware?pullRequest=161)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer-plugin-vmware?pullRequest=161)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**1** out of **2** committers have signed the CLA.

- [x] tenthirtyam
- [ ] niwamo

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer-plugin-vmware?pullRequest=161).</sub>

niwamo commented:
All issues identified by the first round of checks should be addressed. There were three categories of issues:

1. Tests for driver_config.go that I failed to update, including an error message that was intentionally changed, and the adjustment to only set esxi-relevant defaults if esxi is selected; Both have been fixed
2. A small number of linter-identified issues in driver_vmrest.go. All good catches; All now addressed.
3. Exactly two lines with discrepancies between docs and web-docs; Now fixed

I've also re-run my live test of the vmrest builder to ensure changes did not affect functionality

tenthirtyam commented on .web-docs/components/builder/vmx/README.md:
```diff
@@ -1132,7 +1140,65 @@ Since ovftool is only capable of password based authentication
-### SSH key pair automation
+### Building with the VMWare Workstation Pro API (vmrest)
+
+VMWare Workstation Pro ships with a 'vmrest' executable enabling programmatic 
```
"VMware" for all occurrences of "VMWare", please.

niwamo replied:
Thank you for the review! I've corrected "VMWare" to "VMware" in all instances where I wrote the former. 

(Note: I did not make the change for instances of "VMWare" that preexisted my contribution)

niwamo commented:
@tenthirtyam Since you were kind enough to review - do you know if there are additional steps I should be taking and/or reviewers I should be talking to? This has been slow to take off, and I want to make sure I'm doing my due diligence. The members of my team that need this functionality are already using a fork, but it would be nice to have this in the official plugin.

nywilken commented:
@tenthirtyam we will throw some time on your calendar to sync up on this change before starting to review.

niwamo commented:
@tenthirtyam Thanks for the recent follow-up(s). Going to close this merge request, re-factor/clean-up the code, and re-open a request in the next couple of weeks. Will leave the same issue open. Thanks.

tenthirtyam commented:
HI @niwamo 👋 - sorry. I've been meaning to get back with you on this one.

I'm in the midst of doing some consolidation of the drivers based on the recent changes the desktop hypervisors (personal license, too). Those are moving along well and we'll plan to pull out Player around the next major release for Workstation/Fusion and Player EOL.

With regards to `vmrest`, I'd suggest pausing on this one for the time being. There are some product enhancements planned (can't share publically, sorry!) where there will be some changes that may make these changes moot but lead to a more streamlined approach.

**Ryan Johnson**
Distinguished Engineer, VMware by Broadcom
